WIP Purge old data from sqlite db on hass start-up

justyns opened the pull request:
**Description:**

Each time HASS starts, the recorder component will run _purge_old_data() before going into the event loop.   _purge_old_data() will delete records from the `events` and `states` tables older than X days.  After deleting those records, `VACUUM` is run on the database to free up disk space taken by the deleted records.

This also adds a `purge_days` config option to the history component.  By default it is set to `-1` which means no old records will be deleted.

Let me know if `purge_days` should be changed to something else.  I considered `purge_after` and a few other similar names, but figured purge_days would be obvious that it's expecting days.

**Related issue (if applicable):** #1337 

**Example entry for `configuration.yaml` (if applicable):**

``` yaml
# Enables support for tracking state changes over time.
recorder:
  # Delete events and states older than 2 weeks
  purge_days: 14
```

**Checklist:**

If code communicates with devices:
- [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
- [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example](https://github.com/balloob/home-assistant/blob/dev/homeassistant/components/keyboard.py#L16)).
- [ ] New dependencies are only imported inside functions that use them ([example](https://github.com/balloob/home-assistant/blob/dev/homeassistant/components/keyboard.py#L51)).
- [ ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
- [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
- [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
- [ ] Tests have been added to verify that the new code works.

justyns committed:
Add 'purge_days' option to the history/recorder component

Issue https://github.com/balloob/home-assistant/issues/1337

When purge_days is set under the history component, recorder.py will
delete all events and states that are older than purge_days days ago.

Currently, this is only done once at start up.   A vacuum command is
also run to free up the disk space sqlite would still use after deleting
records.

justyns commented:
pylint question:
`R:167, 0: Too many instance attributes (8/7) (too-many-instance-attributes)`

Should I add a comment to ignore this, or is there a recommendation for what to do here?  I add self.config to the Recorder class to get the purge_days config item.

balloob commented on homeassistant/components/recorder.py:
```diff
@@ -168,11 +167,12 @@ def where_limit_to_run(self):
     """A threaded recorder class."""
 
-    def __init__(self, hass):
+    def __init__(self, hass, config):
```
The recorder should not be aware of a config existing. Instead, let's pass down the parameter for how many days to purge (or `None` if not to purge at all)

balloob commented on homeassistant/components/recorder.py:
```diff
@@ -102,14 +102,13 @@ def setup(hass, config):
     global _INSTANCE
-
-    _INSTANCE = Recorder(hass)
+    _INSTANCE = Recorder(hass, config.get('history', {}))
```
Can you add a `CONFIG_SCHEMA` constant to this component to validate and coerce your newly added config option. See [group component](https://github.com/balloob/home-assistant/blob/dev/homeassistant/components/group.py#L61-L63) for an example.

balloob commented:
Yes, you're allowed to ignore that pylint warning.

This will require tests to be written.

justyns committed:
Add CONFIG_SCHEMA to verify config.  Move purge_days key name to
CONF_PURGE_DAYS so it can be changed easier later.

Use 'recorder' domain instead of 'history' domain.

Pass purge_days config directly into Recorder object instead of passing
the config object around.

infamy commented:
@justyns are you still working on adding test cases? Need a hand with that?

justyns commented:
Thanks @infamy .   I started adding tests for this the other day, but ran into a couple issues:

1.  I can modify the config passed to recorder in the `setUp` function, but if I want to test with `purge_days` with different values(off and on), should I just create a new `TestRecorderPurge` class with the different config? I'm not sure if there's a cleaner way to test multiple configs.  
2.  I didn't see an obvious existing way to insert states/events with a custom timestamp.   Should I modify `recorder.record_state` and `recorder.record_event` to accept a date/timestamp parameter, or should I just use a raw sql query to insert a few states manually?  I'm guessing the latter would make sense for testing.

I'm planning on trying to finish this this weekend, but I don't really have anything worth keeping at the moment if anyone wanted to jump in on the testing.

justyns committed:
Add tests for purging old states and events

justyns committed:
Change sqlite queries to work with older versions of sqlite

justyns commented:
@infamy or @balloob, do these tests look okay?   It tests whether `purge_days` is off or set to a number, and verifies it doesn't delete more than it should.  I'll fix the lint errors, and clean it up a little, but wanted to check if there's anything else I should be testing since this is the recorder/history component.

infamy commented:
@justyns you have a bunch of lint errors in the test. (take a look at the Tavis CI output) they should be easy fixes.

As for the test cases those make sense to me. 

The only thing I would add, is a base feature to have it do the clean up every day at midnight, instead of just at startup.

You would need to add the import
```from homeassistant.helpers.event import (track_utc_time_change)```

And create the event to be called 
```track_utc_time_change(hass, lambda now: purge_data(), hour=0)``` 

Or something like that. That would allow HA to do the clean up nightly. (only create the event handle if it is enabled of course)

balloob commented on homeassistant/components/recorder.py:
```diff
@@ -475,6 +486,32 @@ def _close_run(self):
+        _LOGGER.info("Purging events created before %s", purge_before)
+        deleted_rows = self.query(
+            sql_query="DELETE FROM events WHERE created < ?;",
+            data=(int(purge_before.timestamp()),),
```
You should be able to just put in `purge_before` and it will be translated.

balloob commented on homeassistant/components/recorder.py:
```diff
@@ -194,6 +204,7 @@ def run(self):
         """Start processing events to save."""
         self._setup_connection()
         self._setup_run()
+        self._purge_old_data()
```
I think it would make sense to schedule this some time after launch so it won't slow down the launch of Home Assistant.

```python
from homeassistant.helpers.event import track_point_in_utc_time

track_point_in_utc_time(self.hass, lambda now: self._purge_old_data(),
                        dt_util.utcnow() + timedelta(minutes=5)
```

balloob replied:
As this will break your tests, you can just call `recorder._purge_old_data()` in your tests

balloob commented on homeassistant/components/recorder.py:
```diff
@@ -30,6 +31,14 @@
+CONF_PURGE_DAYS = "purge_days"
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.All(dict, {
+        CONF_PURGE_DAYS: int
```
If it's optional, please use `vol.Optional(CONF_PURGE_DAYS)`

balloob commented on homeassistant/components/recorder.py:
```diff
@@ -475,6 +486,32 @@ def _close_run(self):
+    def _purge_old_data(self):
+        """Purge events and states older than purge_days ago."""
+        if not self.purge_days or self.purge_days < 1:
+            _LOGGER.debug("purge_days set to %s, will not purge any old data.",
```
No need for this if you never call it when purge_days is not None, like i suggested above.

balloob commented on homeassistant/components/recorder.py:
```diff
@@ -475,6 +486,32 @@ def _close_run(self):
+
+        purge_before = dt_util.utcnow() - timedelta(days=self.purge_days)
+
+        _LOGGER.info("Purging events created before %s", purge_before)
```
I would merge this with the query below and only print it to `info` if there were actually any rows deleted. Purged X events created before X

balloob commented on homeassistant/components/recorder.py:
```diff
@@ -475,6 +486,32 @@ def _close_run(self):
+            return_value=RETURN_ROWCOUNT)
+        _LOGGER.debug("Deleted %s events", deleted_rows)
+
+        _LOGGER.info("Purging states created before %s", purge_before)
```
Same

balloob commented:
The tests look great 👍  

Added some comments.
